git_machete/cmd.py

```diff
@@ -339,359 +339,1027 @@ def popen_git(cli_ctxt: CommandLineContext, git_cmd: str, *args: str, **kwargs:
-up_branch: Dict[str, str] = {}  # TODO (#110): default dict with None
-indent: Optional[str] = None
-roots: List[str] = []
-annotations: Dict[str, str] = {}
-
+branch_defs_by_sha_in_reflog: Optional[Dict[str, Optional[List[Tuple[str, str]]]]] = None
 
-def expect_in_managed_branches(b: str) -> None:
-    if b not in managed_branches:
-        raise MacheteException(
-            f"Branch `{b}` not found in the tree of branch dependencies.\nUse `git machete add {b}` or `git machete edit`")
+BRANCH_DEF = Tuple[str, str]
+empty_line_status: Optional[bool] = None
 
 
-def expect_at_least_one_managed_branch() -> None:
-    if not roots:
-        raise_no_branches_error()
+class MacheteClient:
 
+    cli_ctxt: CommandLineContext = CommandLineContext()
+    DISCOVER_DEFAULT_FRESH_BRANCH_COUNT = 10
+    PICK_FIRST_ROOT: int = 0
+    PICK_LAST_ROOT: int = -1
+    definition_file_path: str = ""
 
-def raise_no_branches_error() -> None:
-    raise MacheteException(
-        f"No branches listed in {definition_file_path}; use `git machete discover` or `git machete edit`, or edit {definition_file_path} manually.")
-
-
-def read_definition_file(cli_ctxt: CommandLineContext, verify_branches: bool = True) -> None:
-    global indent, managed_branches, down_branches, up_branch, roots, annotations
-
-    with open(definition_file_path) as f:
-        lines: List[str] = [line.rstrip() for line in f.readlines() if not line.isspace()]
-
-    managed_branches = []
-    down_branches = {}
-    up_branch = {}
-    indent = None
-    roots = []
-    annotations = {}
-    at_depth = {}
-    last_depth = -1
-    hint = "Edit the definition file manually with `git machete edit`"
-
-    invalid_branches: List[str] = []
-    for idx, l in enumerate(lines):
-        pfx = "".join(itertools.takewhile(str.isspace, l))
-        if pfx and not indent:
-            indent = pfx
-
-        b_a: List[str] = l.strip().split(" ", 1)
-        b = b_a[0]
-        if len(b_a) > 1:
-            annotations[b] = b_a[1]
-        if b in managed_branches:
-            raise MacheteException(
-                f"{definition_file_path}, line {idx + 1}: branch `{b}` re-appears in the tree definition. {hint}")
-        if verify_branches and b not in local_branches(cli_ctxt):
-            invalid_branches += [b]
-        managed_branches += [b]
-
-        if pfx:
-            depth: int = len(pfx) // len(indent)
-            if pfx != indent * depth:
-                mapping: Dict[str, str] = {" ": "<SPACE>", "\t": "<TAB>"}
-                pfx_expanded: str = "".join(mapping[c] for c in pfx)
-                indent_expanded: str = "".join(mapping[c] for c in indent)
-                raise MacheteException(
-                    f"{definition_file_path}, line {idx + 1}: invalid indent `{pfx_expanded}`, expected a multiply of `{indent_expanded}`. {hint}")
+    def __init__(self, cli_ctxt: CommandLineContext = None) -> None:
```

Comment:
TBH I'd rather not leave the possibility to rely on the defaults... let the users of  `MacheteClient` pass the `CommandLineContext` explicitly

Reply:
Even if that's going to be just a freshly-created `CommandLineContext()`